### PR TITLE
feat(control-panel): media_integrity

### PR DESCRIPTION
**Pull request by punctuations**
adds the commands (fixed and added) to the control panel for being able to remotely change the clock.

Note: the clock will only update upon restart still, as it will have to pull changes on reboot.

### CODE REVIEW

**Comment by punctuations** [comment]
Added the verifying of media integrity as per [5eb1612](https://github.com/timberline-secondary/hackerspace-scripts-2/pull/48/commits/5eb16121f7cc1b8cd20a8346a74dca062a2723fe)

Closing #49; #46 #47

File: scripts/_utils/utils.py [hunk 1]
```diff
@@ -84,6 +91,50 @@ def verify_mimetype(file_url, mimetype_string, local=False):
...
 
+def is_ffmpeg_compatible(file_url) -> bool:
+    command = "ffmpeg -y -v error -i " + file_url + " /tmp/verify-ffmpeg.mp4"
+    err = subprocess.run(command.split(" "), capture_output=True).stderr
+    return err == b''
+
+
+def verify_media_integrity(file_url, mime, local) -> Tuple[bool, Union[str, None], bool]:
+    """
+    Verifies and fixes media
```
Comment (tylerecouture) [inline scripts/_utils/utils.py:102]: How, why?  Details pelase. I'd prefer a novel than just a repeat of the function name

Comment (tylerecouture) [reply scripts/_utils/utils.py:102]: For example, it is only looking for images. what kind of images? all of them? how about animated gifs?  I should not have to read the code to know what this method does, and what it returns.

File: scripts/_utils/utils.py [hunk 2]
```diff
@@ -84,6 +91,50 @@ def verify_mimetype(file_url, mimetype_string, local=False):
...
+    :returns: status and file path and local
+    """
+    try:  # test if input is image
+        im = Image.open(file_url)
+    except PIL.UnidentifiedImageError:  # input is not image
+        return True, file_url, local
+
+    if mime == 'image/gif':
```
Comment (tylerecouture) [inline scripts/_utils/utils.py:110]: This shouldn't be in the verify_media_integrity method, it should be in it's own "process_gif" method

**Review by tylerecouture — CHANGES_REQUESTED** [review]
I left some comments.  Preferably, each issue is its own PR unless they are inextricably linked.

File: scripts/_utils/utils.py (hunk 2 above)
Comment (punctuations) [reply scripts/_utils/utils.py:121]: fixed in [ab06ea3](https://github.com/timberline-secondary/hackerspace-scripts-2/pull/48/commits/ab06ea3ff440c835606d11ff5f6e2405f1d65554)

File: scripts/_utils/utils.py (hunk 1 above)
Comment (punctuations) [reply scripts/_utils/utils.py]: fixed in [ab06ea3](https://github.com/timberline-secondary/hackerspace-scripts-2/pull/48/commits/ab06ea3ff440c835606d11ff5f6e2405f1d65554)

File: scripts/_utils/utils.py [hunk 3]
```diff
@@ -84,6 +91,50 @@ def verify_mimetype(file_url, mimetype_string, local=False):
...
+def is_ffmpeg_compatible(file_url) -> bool:
+    command = "ffmpeg -y -v error -i " + file_url + " /tmp/verify-ffmpeg.mp4"
+    err = subprocess.run(command.split(" "), capture_output=True).stderr
+    return err == b''
+
+
+def verify_media_integrity(file_url, mime, local) -> Tuple[bool, Union[str, None], bool]:
+    """
+    Verifies and fixes media
+    :returns: status and file path and local
```
[the remark this answers is not in the document]
Comment (punctuations) [reply scripts/_utils/utils.py]: fixed in [ab06ea3](https://github.com/timberline-secondary/hackerspace-scripts-2/pull/48/commits/ab06ea3ff440c835606d11ff5f6e2405f1d65554)

File: scripts/Projector/flex_schedule_on.py [hunk 4]
```diff
@@ -5,6 +5,6 @@
```
[the remark this answers is not in the document]
Comment (punctuations) [reply scripts/Projector/flex_schedule_on.py]: fixed as per [4314267](https://github.com/timberline-secondary/hackerspace-scripts-2/pull/48/commits/43142675a3900c1fc130a4a85219f6b39dca2486) & [5132656](https://github.com/timberline-secondary/hackerspace-scripts-2/pull/48/commits/5132656ea97d44293c1fc940cda63bceffc001ce), pull request is #50

File: scripts/Projector/flex_schedule_off.py [hunk 5]
```diff
@@ -5,6 +5,6 @@
```
[the remark this answers is not in the document]
Comment (punctuations) [reply scripts/Projector/flex_schedule_off.py]: fixed as per [4314267](https://github.com/timberline-secondary/hackerspace-scripts-2/pull/48/commits/43142675a3900c1fc130a4a85219f6b39dca2486) & [5132656](https://github.com/timberline-secondary/hackerspace-scripts-2/pull/48/commits/5132656ea97d44293c1fc940cda63bceffc001ce), pull request is #50

File: scripts/Projector/set_flex_schedule.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import json
```
[the remark this answers is not in the document]
Comment (punctuations) [reply scripts/Projector/set_flex_schedule.py]: fixed as per [4314267](https://github.com/timberline-secondary/hackerspace-scripts-2/pull/48/commits/43142675a3900c1fc130a4a85219f6b39dca2486) & [5132656](https://github.com/timberline-secondary/hackerspace-scripts-2/pull/48/commits/5132656ea97d44293c1fc940cda63bceffc001ce), pull request is #50

File: scripts/_utils/utils.py [hunk 7]
```diff
@@ -84,6 +91,54 @@ def verify_mimetype(file_url, mimetype_string, local=False):
         return False
 
 
+def is_ffmpeg_compatible(file_url) -> bool:
```
Comment (tylerecouture) [inline scripts/_utils/utils.py:94]: Will this work with a web url or just a filepath?  No docstring on this method so won't be helpful to developers in the future.  Code looks like it needs a filepath but variable name is url?

File: scripts/_utils/utils.py [hunk 8]
```diff
@@ -84,6 +91,54 @@ def verify_mimetype(file_url, mimetype_string, local=False):
...
+    command = "ffmpeg -y -v error -i " + file_url + " /tmp/verify-ffmpeg.mp4"
+    err = subprocess.run(command.split(" "), capture_output=True).stderr
+    return err == b''
+
+
+def process_gif(im, file_url) -> Tuple[bool, Union[str, None], bool]:
```
Comment (tylerecouture) [inline scripts/_utils/utils.py:100]: Needs method docstring

File: scripts/_utils/utils.py [hunk 9]
```diff
@@ -84,6 +91,54 @@ def verify_mimetype(file_url, mimetype_string, local=False):
...
+        return True, '/tmp/verified.png', True
+    else:  # animated gif -> mp4
+        clip = mp.VideoFileClip(file_url)
+        clip.write_videofile("/tmp/verified.mp4")
+        return True, '/tmp/verified.mp4', True
+
+
+def verify_media_integrity(file_url, mime, local) -> Tuple[bool, Union[str, None], bool]:
+    """
+    Verifies image media integrity (i.e. png, jpg, gif, etc.)
```
[the remark this answers is not in the document]
Comment (tylerecouture) [reply scripts/_utils/utils.py:113]: Does this do anything with animated vs static gifs?

**Review by tylerecouture — CHANGES_REQUESTED** [review]
Mostly need method dosctrings

**Comment by punctuations** [comment]
> @punctuations what exactly does this PR do? The first comment is talking about clocks. Very confused. I need to test this but not of the PR comments/description are detailed enough.

What do you mean?

**Pull request merged**